### PR TITLE
Add `tflint` pre-commit hook

### LINKED ISSUE #41: Enable `tflint` pre-commit hook
\e[0m\e[32mCommand 'tflint --init' successfully done:\e[0m

\e[0m\e[33mTFLint in examples/complete-example/:\e[0m
12 issue(s) found:

Warning: terraform "required_version" attribute is required (terraform_required_version)

  on  line 0:
   (source code not available)

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_version.md

Warning: Interpolation-only expressions are deprecated in Terraform v0.12.14 (terraform_deprecated_interpolation)

  on main.tf line 5:
   5:     Blueprint  = "${replace(basename(path.cwd), "_", "-")}" # tag names based on the directory name

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_deprecated_interpolation.md

Warning: Interpolation-only expressions are deprecated in Terraform v0.12.14 (terraform_deprecated_interpolation)

  on main.tf line 48:
  48:       username = "${var.aws_admin_1_username}"

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_deprecated_interpolation.md

Warning: Interpolation-only expressions are deprecated in Terraform v0.12.14 (terraform_deprecated_interpolation)

  on main.tf line 53:
  53:       username = "${var.aws_admin_2_username}"

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_deprecated_interpolation.md

Warning: data "aws_eks_cluster_auth" "this" is declared but not used (terraform_unused_declarations)

  on providers.tf line 2:
   2: data "aws_eks_cluster_auth" "this" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: Missing version constraint for provider "aws" in "required_providers" (terraform_required_providers)

  on providers.tf line 14:
  14: provider "aws" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_providers.md

Warning: Missing version constraint for provider "kubernetes" in "required_providers" (terraform_required_providers)

  on providers.tf line 19:
  19: provider "kubernetes" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_providers.md

Warning: Missing version constraint for provider "helm" in "required_providers" (terraform_required_providers)

  on providers.tf line 29:
  29: provider "helm" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_providers.md

Warning: variable "aws_profile" is declared but not used (terraform_unused_declarations)

  on variables.tf line 19:
  19: variable "aws_profile" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "create_database_subnet_group" is declared but not used (terraform_unused_declarations)

  on variables.tf line 48:
  48: variable "create_database_subnet_group" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "create_database_subnet_route_table" is declared but not used (terraform_unused_declarations)

  on variables.tf line 54:
  54: variable "create_database_subnet_route_table" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "assign_public_ip" is declared but not used (terraform_unused_declarations)

  on variables.tf line 84:
  84: variable "assign_public_ip" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

\e[0m\e[33mTFLint in examples/tf-state-backend/:\e[0m
2 issue(s) found:

Warning: terraform "required_version" attribute is required (terraform_required_version)

  on  line 0:
   (source code not available)

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_version.md

Warning: Missing version constraint for provider "aws" in "required_providers" (terraform_required_providers)

  on main.tf line 5:
   5: data "aws_partition" "current" {}

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_providers.md

\e[0m\e[33mTFLint in modules/bastion/:\e[0m
28 issue(s) found:

Warning: local.root_volume_config is declared but not used (terraform_unused_declarations)

  on locals.tf line 32:
  32:   root_volume_config = {
  33:     volume_type = "gp3"
  34:     volume_size = "20"
  35:     encrypted   = true
  36:   }

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: data "aws_region" "current" is declared but not used (terraform_unused_declarations)

  on main.tf line 3:
   3: data "aws_region" "current" {}

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: List items should be accessed using square brackets (terraform_deprecated_index)

  on main.tf line 37:
  37:   vpc_security_group_ids      = length(local.security_group_configs) > 0 ? aws_security_group.sg.*.id : var.security_group_ids

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_deprecated_index.md

Warning: Missing version constraint for provider "tls" in "required_providers" (terraform_required_providers)

  on main.tf line 60:
  60: resource "tls_private_key" "bastion_key" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_providers.md

Warning: Missing version constraint for provider "cloudinit" in "required_providers" (terraform_required_providers)

  on main.tf line 185:
 185: data "cloudinit_config" "config" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_providers.md

Warning: List items should be accessed using square brackets (terraform_deprecated_index)

  on output.tf line 21:
  21:   value       = length(local.security_group_configs) > 0 ? aws_security_group.sg.*.id : var.security_group_ids

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_deprecated_index.md

Warning: variable "ec2_key_name" is declared but not used (terraform_unused_declarations)

  on variables.tf line 56:
  56: variable "ec2_key_name" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "requires_eip" is declared but not used (terraform_unused_declarations)

  on variables.tf line 80:
  80: variable "requires_eip" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "user_data" is declared but not used (terraform_unused_declarations)

  on variables.tf line 86:
  86: variable "user_data" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "role_name" is declared but not used (terraform_unused_declarations)

  on variables.tf line 92:
  92: variable "role_name" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "acl" is declared but not used (terraform_unused_declarations)

  on variables.tf line 172:
 172: variable "acl" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "versioning_enabled" is declared but not used (terraform_unused_declarations)

  on variables.tf line 190:
 190: variable "versioning_enabled" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "force_destroy" is declared but not used (terraform_unused_declarations)

  on variables.tf line 196:
 196: variable "force_destroy" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "logging" is declared but not used (terraform_unused_declarations)

  on variables.tf line 202:
 202: variable "logging" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "bucket_public_access_block" is declared but not used (terraform_unused_declarations)

  on variables.tf line 220:
 220: variable "bucket_public_access_block" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: `ssh_user` variable has no type (terraform_typed_variables)

  on variables.tf line 230:
 230: variable "ssh_user" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_typed_variables.md

Warning: `enable_hourly_cron_updates` variable has no type (terraform_typed_variables)

  on variables.tf line 234:
 234: variable "enable_hourly_cron_updates" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_typed_variables.md

Warning: variable "enable_hourly_cron_updates" is declared but not used (terraform_unused_declarations)

  on variables.tf line 234:
 234: variable "enable_hourly_cron_updates" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: `keys_update_frequency` variable has no type (terraform_typed_variables)

  on variables.tf line 238:
 238: variable "keys_update_frequency" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_typed_variables.md

Warning: variable "keys_update_frequency" is declared but not used (terraform_unused_declarations)

  on variables.tf line 238:
 238: variable "keys_update_frequency" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: `user_data_file` variable has no type (terraform_typed_variables)

  on variables.tf line 242:
 242: variable "user_data_file" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_typed_variables.md

Warning: variable "user_data_file" is declared but not used (terraform_unused_declarations)

  on variables.tf line 242:
 242: variable "user_data_file" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: `additional_user_data_script` variable has no type (terraform_typed_variables)

  on variables.tf line 246:
 246: variable "additional_user_data_script" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_typed_variables.md

Warning: variable "ssh_public_key_names" is declared but not used (terraform_unused_declarations)

  on variables.tf line 250:
 250: variable "ssh_public_key_names" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: `cluster_sops_policy_arn` variable has no type (terraform_typed_variables)

  on variables.tf line 255:
 255: variable "cluster_sops_policy_arn" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_typed_variables.md

Warning: `ssmkey_arn` variable has no type (terraform_typed_variables)

  on variables.tf line 277:
 277: variable "ssmkey_arn" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_typed_variables.md

Warning: variable "ssmkey_arn" is declared but not used (terraform_unused_declarations)

  on variables.tf line 277:
 277: variable "ssmkey_arn" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "vpc_endpoints_enabled" is declared but not used (terraform_unused_declarations)

  on variables.tf line 324:
 324: variable "vpc_endpoints_enabled" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

\e[0m\e[33mTFLint in modules/eks/:\e[0m
5 issue(s) found:

Warning: local.azs is declared but not used (terraform_unused_declarations)

  on locals.tf line 5:
   5:   azs = slice(data.aws_availability_zones.available.names, 0, 3)

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: data "aws_eks_cluster_auth" "this" is declared but not used (terraform_unused_declarations)

  on main.tf line 1:
   1: data "aws_eks_cluster_auth" "this" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)

  on main.tf line 322:
 322:                "AWS": ${var.cluster_kms_key_additional_admin_arns == [] ? "[]" : jsonencode(var.cluster_kms_key_additional_admin_arns)}

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_empty_list_equality.md

Warning: variable "public_subnet_ids" is declared but not used (terraform_unused_declarations)

  on variables.tf line 26:
  26: variable "public_subnet_ids" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "aws_account" is declared but not used (terraform_unused_declarations)

  on variables.tf line 37:
  37: variable "aws_account" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

\e[0m\e[33mTFLint in modules/rds/:\e[0m
2 issue(s) found:

Warning: variable "db_vpc_security_group_ids" is declared but not used (terraform_unused_declarations)

  on variables.tf line 13:
  13: variable "db_vpc_security_group_ids" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "create_db_subnet_group" is declared but not used (terraform_unused_declarations)

  on variables.tf line 85:
  85: variable "create_db_subnet_group" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

\e[0m\e[33mTFLint in modules/s3-irsa/:\e[0m
1 issue(s) found:

Warning: local.name is declared but not used (terraform_unused_declarations)

  on main.tf line 2:
   2:   name                = basename(path.cwd)

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

\e[0m\e[33mTFLint in modules/sops/:\e[0m
2 issue(s) found:

Warning: local.name is declared but not used (terraform_unused_declarations)

  on main.tf line 2:
   2:   name                = basename(path.cwd)

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "region" is declared but not used (terraform_unused_declarations)

  on variables.tf line 1:
   1: variable "region" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

\e[0m\e[33mTFLint in modules/tfstate-backend/:\e[0m
9 issue(s) found:

Warning: terraform "required_version" attribute is required (terraform_required_version)

  on  line 0:
   (source code not available)

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_version.md

Warning: local.role_name is declared but not used (terraform_unused_declarations)

  on locals.tf line 2:
   2:   role_name     = "${var.bucket_prefix}-role"

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: local.bucket_prefix is declared but not used (terraform_unused_declarations)

  on locals.tf line 3:
   3:   bucket_prefix = var.bucket_prefix

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: data "aws_partition" "current" is declared but not used (terraform_unused_declarations)

  on main.tf line 1:
   1: data "aws_partition" "current" {}

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: Missing version constraint for provider "aws" in "required_providers" (terraform_required_providers)

  on main.tf line 66:
  66: resource "aws_s3_bucket_policy" "backend_bucket" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_providers.md

Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)

  on main.tf line 76:
  76:         "AWS": ${var.cluster_key_admin_arns == [] ? "[]" : jsonencode(var.cluster_key_admin_arns)}

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_empty_list_equality.md

Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)

  on main.tf line 95:
  95:         "AWS": ${var.cluster_key_admin_arns == [] ? "[]" : jsonencode(var.cluster_key_admin_arns)}

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_empty_list_equality.md

Warning: variable "region" is declared but not used (terraform_unused_declarations)

  on variables.tf line 1:
   1: variable "region" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "permissions_boundary" is declared but not used (terraform_unused_declarations)

  on variables.tf line 28:
  28: variable "permissions_boundary" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

\e[0m\e[33mTFLint in modules/vpc/:\e[0m
4 issue(s) found:

Warning: data "aws_iam_policy_document" "dynamodb_endpoint_policy" is declared but not used (terraform_unused_declarations)

  on main.tf line 196:
 196: data "aws_iam_policy_document" "dynamodb_endpoint_policy" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: data "aws_iam_policy_document" "generic_endpoint_policy" is declared but not used (terraform_unused_declarations)

  on main.tf line 216:
 216: data "aws_iam_policy_document" "generic_endpoint_policy" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "region" is declared but not used (terraform_unused_declarations)

  on variables.tf line 1:
   1: variable "region" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md

Warning: variable "create_database_subnet_route_table" is declared but not used (terraform_unused_declarations)

  on variables.tf line 39:
  39: variable "create_database_subnet_route_table" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_unused_declarations.md
```

### PR DESCRIPTION
**Pull request by RothAndrew**
Just uses the `terraform` plugin for now. Will make a new issue to turn on the `aws` plugin.

Also does a bunch of cleanup to the nodegroup configs, which makes the code more maintainable and saves us money. The stuff that was being used previously was more expensive.

Closes #41

### CODE REVIEW

File: examples/complete/main.tf [hunk 1]
```diff
@@ -207,18 +201,6 @@ module "eks" {
...
-    instance_types = ["m6i.large", "m5.large", "m5n.large", "m5zn.large"]
-
-    # We are using the IRSA created below for permissions
-    # However, we have to deploy with the policy attached FIRST (when creating a fresh cluster)
-    # and then turn this off after the cluster/node group is created. Without this initial policy,
-    # the VPC CNI fails to assign IPs and nodes cannot join the cluster
-    # See https://github.com/aws/containers-roadmap/issues/1666 for more context
-    iam_role_attach_cni_policy = true
-  }
-
```
Comment (zack-is-cool) [inline examples/complete/main.tf:221]: recommend adding this back in

Comment (RothAndrew) [reply examples/complete/main.tf:221]: Do we need it? It turned out to have not been used, since that variable was never wired into the upstream module.

Comment (zack-is-cool) [reply examples/complete/main.tf:221]: No, I guess not. At least it will be available (upstream eventually) if someone wants to use it.

**Review by zack-is-cool — DISMISSED** [review]
LGTM

File: examples/complete/main.tf [hunk 2]
```diff
@@ -25,49 +23,76 @@ locals {
     }
   )
 
+  eks_managed_node_group_defaults = {
+    # https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/node_groups.tf
+    create                        = var.enable_eks_managed_nodegroups
```
Comment (zack-is-cool) [inline examples/complete/main.tf:28]: this won't work

File: examples/complete/main.tf [hunk 3]
```diff
@@ -25,49 +23,76 @@ locals {
...
-
-      # ami_id = "" # defaults to latest amazon linux 2 eks ami matching k8s version in the upstream module
-      # create_iam_role           = true                                                    # Changing `create_iam_role=false` to bring your own IAM Role
-      # iam_role_arn              = module.eks.aws_iam_role_self_managed_ng_arn              # custom IAM role for aws-auth mapping; used when create_iam_role = false
-      # iam_instance_profile_name = module.eks.aws_iam_instance_profile_self_managed_ng_name # IAM instance profile name for Launch templates; used when create_iam_role = false
-      placement = {
-        tenancy = var.eks_worker_tenancy
-      }
+  self_managed_node_group_defaults = {
+    create                                 = var.enable_self_managed_nodegroups
```
Comment (zack-is-cool) [inline examples/complete/main.tf:44]: this won't work

File: examples/complete/main.tf (hunk 2 above)
Comment (RothAndrew) [reply examples/complete/main.tf:28]: ah yeah you're right. Looks like they don't look at the defaults for that particular field. which is.. kinda dumb. You should be able to put `create` in the defaults

**Review by zack-is-cool — APPROVED** [review]
LGTM (if pass test, if not, LBTM)

**Pull request merged**